### ISSUE TITLE
[BUG] KeyVault ImportKey() gives "Service request failed" Error 400 when importing a key with invalid name

### ISSUE THREAD

**Issue by mnns** [Azure/azure-sdk-for-net#15784]
**Describe the bug**
I'm getting Error 400 when trying to import an ECDsa key using ImportKeyAsync(), struggled to find out the actual problem, then found out I passed an invalid key name.

**Expected behavior**
Importing an ECDsa key.

**Actual behavior (include Exception or Stack Trace)**
`Runtime error
Service request failed.
Status: 400 (Bad Request)`

**To Reproduce**
Simply use "ImportKeyAsync()" with the key name: "Test: my organization (13948)" and it will crash.

**Environment:**
Azure.Security.KeyVaults.Keys.KeyClient, .NET Core 3.1, Windows 10

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team member best able to assist.

**Comment by heaths** [comment]
Normally, the exception message describes that the name is invalid. By [design](https://azure.github.io/azure-sdk/general_implementation.html#parameter-validation), we do minimal client-side validation to avoid tight coupling and potentially blocking new service-side behaviors. If I simply use `test secret` as a value, I see this:

```bash
$ dotnet run -- 'test secret'
Unhandled exception. Azure.RequestFailedException: Service request failed.
Status: 400 (Bad Request)

Content:
{"error":{"code":"BadParameter","message":"The request URI contains an invalid name: test secret"}}
```

However, in this particular case with the example you gave typed exactly as is, I get an ASP.NET error page back:

```
Unhandled exception. Azure.RequestFailedException: Service request failed.
Status: 400 (Bad Request)

Content:
<!DOCTYPE html>
<html>
    <head>
        <title>Runtime Error</title>
        <meta name="viewport" content="width=device-width" />
        <style>
         body {font-family:"Verdana";font-weight:normal;font-size: .7em;color:black;}
         p {font-family:"Verdana";font-weight:normal;color:black;margin-top: -5px}
         b {font-family:"Verdana";font-weight:bold;color:black;margin-top: -5px}
         H1 { font-family:"Verdana";font-weight:normal;font-size:18pt;color:red }
         H2 { font-family:"Verdana";font-weight:normal;font-size:14pt;color:maroon }
         pre {font-family:"Consolas","Lucida Console",Monospace;font-size:11pt;margin:0;padding:0.5em;line-height:14pt}
         .marker {font-weight: bold; color: black;text-decoration: none;}
         .version {color: gray;}
         .error {margin-bottom: 10px;}
         .expandable { text-decoration:underline; font-weight:bold; color:navy; cursor:hand; }
         @media screen and (max-width: 639px) {
          pre { width: 440px; overflow: auto; white-space: pre-wrap; word-wrap: break-word; }
         }
         @media screen and (max-width: 479px) {
          pre { width: 280px; }
         }
        </style>
    </head>

    <body bgcolor="white">

            <span><H1>Server Error in '/' Application.<hr width=100% size=1 color=silver></H1>

            <h2> <i>Runtime Error</i> </h2></span>

            <font face="Arial, Helvetica, Geneva, SunSans-Regular, sans-serif ">

            <b> Description: </b>An application error occurred on the server. The current custom error settings for this application prevent the details of the application error from being viewed remotely (for security reasons). It could, however, be viewed by browsers running on the local server machine.
            <br><br>

            <b>Details:</b> To enable the details of this specific error message to be viewable on remote machines, please create a &lt;customErrors&gt; tag within a &quot;web.config&quot; configuration file located in the root directory of the current web application. This &lt;customErrors&gt; tag should then have its &quot;mode&quot; attribute set to &quot;Off&quot;.<br><br>

            <table width=100% bgcolor="#ffffcc">
               <tr>
                  <td>
                      <code><pre>

&lt;!-- Web.Config Configuration File --&gt;

&lt;configuration&gt;
    &lt;system.web&gt;
        &lt;customErrors mode=&quot;Off&quot;/&gt;
    &lt;/system.web&gt;
&lt;/configuration&gt;</pre></code>

                  </td>
               </tr>
            </table>

            <br>

            <b>Notes:</b> The current error page you are seeing can be replaced by a custom error page by modifying the &quot;defaultRedirect&quot; attribute of the application&#39;s &lt;customErrors&gt; configuration tag to point to a custom error page URL.<br><br>

            <table width=100% bgcolor="#ffffcc">
               <tr>
                  <td>
                      <code><pre>

&lt;!-- Web.Config Configuration File --&gt;

&lt;configuration&gt;
    &lt;system.web&gt;
        &lt;customErrors mode=&quot;RemoteOnly&quot; defaultRedirect=&quot;mycustompage.htm&quot;/&gt;
    &lt;/system.web&gt;
&lt;/configuration&gt;</pre></code>

                  </td>
               </tr>
            </table>

            <br>

    </body>
</html>
```

@RandalliLama @jlichwa this seems like a service issue.

**Comment by AshokCMSFT** [comment]
The error is from the web server because the url is not properly encoded. Characters like space, <, > need to be encoded otherwise webserver will reject the request. Client should use UrlEncode. https://docs.microsoft.com/en-us/dotnet/api/system.web.httputility.urlencode?view=netcore-3.1

**Comment by heaths** [comment]
Seems this was caused way back in 7cb3f586423dbe36ed262dbdafd06ff0a5c8cbcc. Investigating why...

**Comment by heaths** [comment]
Actually, that commit just maintained existing behavior, so this was always a problem. Working on a fix...

**Comment by heaths** [comment]
@AshokCMSFT `Uri.EscapeUriString` is actually recommended since that doesn't add a dependency to System.Web to client apps. That said, I still get the error even when spaces are encoded:

Path: https://heathskv.vault.azure.net/keys/Test:%20bad%20name%20(396245)?api-version=7.1
Error: same ASP.NET HTML output as above. Those are valid characters for a URI: https://tools.ietf.org/html/rfc3986#page-22 (a path char can be ":", "(", or ")".

To note, it's not the space causing an issue. I get a proper JSON error response with "bad name", thus there's practically no difference between "Test: bad name (123456)" and "Test:%20bad%20name%20(123456)".

The server response should also contain the valid set of characters to inform callers what is valid. The service response with just a space reads,

> The request URI contains an invalid name: bad name

I.e.

```json
{"error":{"code":"BadParameter","message":"The request URI contains an invalid name: bad name"}}
```

**Comment by heaths** [comment]
Testing `HttpUtility.UrlEncode` vs `Uri.EscapeUriString`, the difference seems to be the former encodes ":" as "%3a", but also " " (space) as "+", which is intended for URI query strings (i.e. equivalent to `Uri.EscapeDataString`). Same for "/" to "%2f", which is a valid path character. If the {name} is part of the path, it should probably be URI path-encoded, not data-encoded (and decoded suitably).

**Comment by sebansal** [comment]
Hi @heaths 
I was triaging this list, can you share if it has been resolved and can be closed?
Thanks

**Comment by heaths** [comment]
It's still an open issue. Our SDK architects aren't crazy about tightly coupling the SDK to current swagger since it changes. Ideally, the service should send back a suitable error that describes what the user should do. In this case, the default ASP.NET error page (HTML) is returned, which isn't a good customer UX. Ideally, this should be fixed by the service.

**Comment by heaths** [comment]
@AshokCMSFT even encoding the name doesn't solve the issue with ":". I don't see this issue with some other services I've tested. Is there nothing that can be fixed server-side so you at least return HTTP 400 instead of an ASP.NET error page?

**Comment by AshokCMSFT** [comment]
I see that we are blocked by IIS. https://blogs.iis.net/wadeh/how-iis-blocks-characters-in-urls. I have not found a good workaround. Services not running on IIS may not have this issue.

**Comment by heaths** [comment]
Architects have given conditional approval that, if we do validate parameters (for example, given your simple character set - we'd probably not use regex for perf of such a simple charset), we need to know that you will *never* deviate from that charset in future iterations. At least, we could also just err if the name includes ":", but this was less palatable because the network call isn't that much slower and potentially more data comes back, i.e. it's an error either way; maybe some day hosting changes returns the right error so working around it now potentially drops data.

As such, I'm inclined to won't fix this but want service team feedback. The error message isn't great (well, not useful at all) but the current behavior is correct (error) and future behavior *might* improve unless we make code changes now and lose out on potential improvements.

**Comment by sebansal** [comment]
We will look for opportunity to fix it in future. With current architecture, it would not be fixable.
